BWIndex.py

```diff
@@ -0,0 +1,239 @@
+		if genome is not None:
+			self.Text = genome
+		elif genomefile is not None:
+			self.loadGenomeFromFile(genomefile)
+		else:
+			print("You must supply a genome file or a genome string.")
+			sys.exit()
+		if self.Text[-1] != '$':
+			print("Genome must be terminated by $")
+			sys.exit()
+		self.buildTree()
+		self.Solve()
+		self.Count = dict()
+		self.letters = {i for i in self.BWT}
+		self.alphabet = [i for i in self.letters]
+		self.alphabet.sort()
+		self.computeCountArray()
+
+		self.FirstOccurrence = dict()
+		freq = dict()
+		for i in self.BWT:
+			freq[i] = freq.get(i, 0) + 1
+		index = 0
+		self.FirstOccurrence['$'] = 0
+		for j in range(1, len(self.alphabet)):
+			prevLetter = self.alphabet[j - 1]
+			letter = self.alphabet[j]
+			index += freq[prevLetter]
+			self.FirstOccurrence[letter] = index
+
+	def buildTree(self):
+		self.G = {0: []}  # parent to child, int -> list
+		self.H = dict()  # child to parent, int -> int
+		self.nodeCount = 1
+		self.NS = dict()  # node start index
+		self.NE = dict()  # node end index
+		self.SS = dict()  # suffix start index
+		for ix in range(len(self.Text)):
+			# print(ix)
+			self.threadSuffix(ix)
+
+	def exportIndex(self, filename, SA_gap=None, C_gap=None):
+		g = open(filename, 'w')
+		self.SA_gap = SA_gap
+		self.C_gap = C_gap
+		g.write(self.BWT + '\n')
+		if SA_gap is None:
+			g.write(','.join(str(i) + ';' + str(s) for i, s in enumerate(self.suffixArray)) + '\n')
+		else:
+			g.write(','.join(str(i) + ';' + str(s) for i, s in enumerate(self.suffixArray) if s % SA_gap == 0) + '\n')
+		g.write(','.join(str(x) for x in sorted(self.Count.keys())) + '\n')
+		for x in sorted(self.Count.keys()):
+			if C_gap is None:
+				g.write(','.join(str(self.Count[x][i]) for i in range(len(self.Count[x]))) + '\n')
+			else:
+				g.write(','.join(str(self.Count[x][i]) for i in range(len(self.Count[x])) if i % C_gap == 0) + '\n')
+		g.write(','.join(str(self.FirstOccurrence[x]) for x in sorted(self.FirstOccurrence.keys())) + '\n')
+		g.write(str(SA_gap) + '\n')
+		g.write(str(C_gap) + '\n')
+
+	def Solve(self):
+		S = [('', 0)]
+		transform = ''
+		suffixArray = []
+		while S:
+			node = S.pop()
+			nodeIx = node[1]
+			if self.G.get(nodeIx) is not None:
+				if self.G[nodeIx] != []:
+					children = [(self.Text[self.NS[child]], child) for child in self.G[nodeIx]]
+					children.sort(reverse=True)
+					S += children
+			else:
+				transform += self.Text[self.SS[nodeIx] - 1]
+				suffixArray.append(self.SS[nodeIx])
+		self.BWT = transform
+		self.suffixArray = suffixArray
+
+	def drawGraph(self, maxLabelLength=None, outFile=None):
+		D = Digraph()
+		self.NS[0] = 0
+		self.NE[0] = -1
+		for i in range(self.nodeCount):
+			label = self.Text[self.NS[i]:self.NE[i] + 1]
+			if maxLabelLength:
+				label = label[:maxLabelLength]
+			D.node(str(i), label)
+		for tail in self.G:
+			for head in self.G[tail]:
+				D.edge(str(tail), str(head))
+		D.render(outFile, view=True)
+
+	def report(self):

```

Comment:
It looks like you're providing a way to summarize or print out an instance of the class, which is a very good and helpful idea. Python comes with two "hooks" for that sort of thing: `__repr__` and `__str__` methods, both of which are to be overloaded and return strings. `__str__` is more for printing out an instance for humans, and the idea behind  `__repr__` is that it should print out something that could be used to completely programmatically recreate the instance. More [here](https://www.geeksforgeeks.org/str-vs-repr-in-python/).